File: src/agenteval/scorer.py
"""Scorer — evaluates agent output across multiple dimensions."""

import subprocess
from pathlib import Path

import anthropic
from rich.console import Console

from agenteval.models import RunResult, ScoreResult, TaskDefinition

console = Console()

LLM_JUDGE_PROMPT = """\
You are an expert code reviewer evaluating a coding agent's work.

## Task
{prompt}

## Expected Solution (diff)
{expected_diff}

## Agent's Solution (diff)
{actual_diff}

Rate the agent's solution on a scale of 1-5:
1 = Completely wrong or no meaningful attempt
2 = Partially addresses the task but major issues
3 = Reasonable attempt, some issues
4 = Good solution, minor issues
5 = Excellent solution, correct and clean

Respond with EXACTLY this format:
SCORE: <number>
EXPLANATION: <one paragraph>
"""


def diff_similarity(expected: str, actual: str) -> float:
    """Jaccard similarity of changed lines between two diffs."""
    def _extract_changed_lines(diff_text: str) -> set[str]:
        lines = set()
        for line in diff_text.splitlines():
            if line.startswith(("+", "-")) and not line.startswith(("+++", "---")):
                lines.add(line.strip())
        return lines

    expected_lines = _extract_changed_lines(expected)
    actual_lines = _extract_changed_lines(actual)

    if not expected_lines and not actual_lines:
        return 1.0
    if not expected_lines or not actual_lines:
        return 0.0

    intersection = expected_lines & actual_lines
    union = expected_lines | actual_lines
    return len(intersection) / len(union)


def _detect_test_commands(repo_path: Path) -> list[str]:
    """Detect test commands based on repo contents."""
    if (repo_path / "pytest.ini").exists() or (repo_path / "pyproject.toml").exists():
        return ["python -m pytest"]
    if (repo_path / "package.json").exists():
        return ["npm test"]
    if (repo_path / "Makefile").exists():
        return ["make test"]
    return []


def _detect_lint_commands(repo_path: Path, changed_files: list[str]) -> list[str]:
    """Detect linter commands based on repo contents."""
    py_files = [f for f in changed_files if f.endswith(".py")]
    js_files = [f for f in changed_files if f.endswith((".js", ".ts", ".jsx", ".tsx"))]

    commands = []
    if py_files:
        file_args = " ".join(py_files)
        commands.append(f"ruff check {file_args}")
    if js_files:
        file_args = " ".join(js_files)
        commands.append(f"eslint {file_args}")
    return commands


class Scorer:
    """Scores agent run results against task definitions."""

    def __init__(self, no_llm: bool = False):
        self.no_llm = no_llm
        if not no_llm:
            self.client = anthropic.Anthropic()

    def score_run(
        self, run: RunResult, task: TaskDefinition, repo_path: Path,
    ) -> ScoreResult:
        """Score a single run result against a task definition."""
        # Load expected diff
        expected_diff = ""
        if task.expected_diff_path and Path(task.expected_diff_path).exists():
            expected_diff = Path(task.expected_diff_path).read_text()

        # Diff similarity
        similarity = diff_similarity(expected_diff, run.actual_diff)

        # Tests
        tests_pass, test_output = self._run_tests(run, task, repo_path)

        # Lint
        lint_clean, lint_output = self._run_lint(run, task, repo_path)

        # LLM judge
        llm_score, llm_explanation = self._llm_judge(run, task, expected_diff)

        # Overall score (weighted composite)
        overall = self._compute_overall(
            tests_pass=tests_pass,
            lint_clean=lint_clean,
            diff_similarity=similarity,
            llm_judge_score=llm_score,
        )

        return ScoreResult(
            task_id=run.task_id,
            agent=run.agent,
            tests_pass=tests_pass,
            test_output=test_output,
            lint_clean=lint_clean,
            lint_output=lint_output,
            diff_similarity=similarity,
            llm_judge_score=llm_score,
            llm_judge_explanation=llm_explanation,
            overall_score=overall,
        )

    def _run_tests(
        self, run: RunResult, task: TaskDefinition, repo_path: Path,
    ) -> tuple[bool | None, str]:
        """Run test commands and return (pass, output)."""
        commands = task.test_commands or _detect_test_commands(repo_path)
        if not commands:
            return None, ""

        try:
            # Checkout base, apply diff
            subprocess.run(
                ["git", "checkout", task.base_commit],
                capture_output=True, text=True, check=True,
                cwd=str(repo_path), timeout=30,
            )
            if run.actual_diff:
                proc = subprocess.run(
                    ["git", "apply", "--allow-empty"],
                    input=run.actual_diff,
                    capture_output=True, text=True,
                    cwd=str(repo_path), timeout=30,
                )
                if proc.returncode != 0:
                    return False, f"Failed to apply diff: {proc.stderr}"

            # Run tests
            outputs = []
            all_pass = True
            for cmd in commands:
                result = subprocess.run(
                    cmd, shell=True, capture_output=True, text=True,
                    cwd=str(repo_path), timeout=120,
                )
                outputs.append(result.stdout + result.stderr)
                if result.returncode != 0:
                    all_pass = False

            return all_pass, "\n".join(outputs)
        except Exception as e:
            return False, str(e)

    def _run_lint(
        self, run: RunResult, task: TaskDefinition, repo_path: Path,
    ) -> tuple[bool | None, str]:
        """Run linter on changed files."""
        changed_files = task.metadata.get("files_changed", [])
        commands = _detect_lint_commands(repo_path, changed_files)
        if not commands:
            return None, ""

        try:
            outputs = []
            all_clean = True
            for cmd in commands:
                result = subprocess.run(
                    cmd, shell=True, capture_output=True, text=True,
                    cwd=str(repo_path), timeout=60,
                )
                outputs.append(result.stdout + result.stderr)
                if result.returncode != 0:
                    all_clean = False
            return all_clean, "\n".join(outputs)
        except Exception as e:
            return False, str(e)

    def _llm_judge(
        self, run: RunResult, task: TaskDefinition, expected_diff: str,
    ) -> tuple[int | None, str]:
        """Use LLM to judge quality. Returns (score 1-5, explanation)."""
        if self.no_llm:
            return None, ""

        prompt = LLM_JUDGE_PROMPT.format(
            prompt=task.prompt,
            expected_diff=expected_diff or "(not available)",
            actual_diff=run.actual_diff or "(empty — agent produced no changes)",
        )

        try:
            response = self.client.messages.create(
                model="claude-sonnet-4-20250514",
                max_tokens=512,
                messages=[{"role": "user", "content": prompt}],
            )
            text = response.content[0].text

            # Parse response
            score = None
            explanation = ""
            for line in text.splitlines():
                if line.startswith("SCORE:"):
                    try:
                        score = int(line.split(":")[1].strip())
                        score = max(1, min(5, score))
                    except ValueError:
                        pass
                elif line.startswith("EXPLANATION:"):
                    explanation = line.split(":", 1)[1].strip()

            return score, explanation
        except Exception as e:
            console.print(f"[yellow]LLM judge failed: {e}[/yellow]")
            return None, str(e)

    @staticmethod
    def _compute_overall(
        tests_pass: bool | None,
        lint_clean: bool | None,
        diff_similarity: float,
        llm_judge_score: int | None,
    ) -> float:
        """Compute weighted overall score (0-1)."""
        total_weight = 0.0
        weighted_sum = 0.0

        if tests_pass is not None:
            total_weight += 0.35
            weighted_sum += 0.35 * (1.0 if tests_pass else 0.0)

        if lint_clean is not None:
            total_weight += 0.1
            weighted_sum += 0.1 * (1.0 if lint_clean else 0.0)

        # Diff similarity always contributes
        total_weight += 0.25
        weighted_sum += 0.25 * diff_similarity

        if llm_judge_score is not None:
            total_weight += 0.3
            weighted_sum += 0.3 * (llm_judge_score / 5.0)

        if total_weight == 0:
            return 0.0
        return round(weighted_sum / total_weight, 4)

File: internal/scorer/scorer.go
// Package scorer evaluates agent run results across multiple dimensions.
package scorer

import (
	"context"
	"fmt"
	"os"
	"os/exec"
	"path/filepath"
	"regexp"
	"strconv"
	"strings"

	"github.com/anthropics/anthropic-sdk-go"

	"agenteval/internal/models"
)

const judgeModel = "claude-sonnet-4-20250514"

const llmJudgePrompt = `You are an expert code reviewer evaluating a coding agent's work.

## Task
%s

## Expected Solution (diff)
%s

## Agent's Solution (diff)
%s

Rate the agent's solution on a scale of 1-5:
1 = Completely wrong or no meaningful attempt
2 = Partially addresses the task but major issues
3 = Reasonable attempt, some issues
4 = Good solution, minor issues
5 = Excellent solution, correct and clean

Respond with EXACTLY this format:
SCORE: <number>
EXPLANATION: <one paragraph>`

// Scorer evaluates agent runs. When NoLLM is true the LLM judge dimension is skipped.
type Scorer struct {
	NoLLM  bool
	client *anthropic.Client
}

// New creates a Scorer. If noLLM is true no Anthropic client is created.
func New(noLLM bool) *Scorer {
	s := &Scorer{NoLLM: noLLM}
	if !noLLM {
		s.client = anthropic.NewClient()
	}
	return s
}

// ScoreRun evaluates a single RunResult against its TaskDefinition.
// repoPath is the local repo path (used for test/lint commands).
func (s *Scorer) ScoreRun(run models.RunResult, task models.TaskDefinition, repoPath string) models.ScoreResult {
	expectedDiff := ""
	if task.ExpectedDiffPath != "" {
		if data, err := os.ReadFile(task.ExpectedDiffPath); err == nil {
			expectedDiff = string(data)
		}
	}

	similarity := DiffSimilarity(expectedDiff, run.ActualDiff)
	testsPass, testOutput := s.runTests(run, task, repoPath)
	lintClean, lintOutput := s.runLint(task, repoPath)
	llmScore, llmExplanation := s.llmJudge(run, task, expectedDiff)

	overall := computeOverall(testsPass, lintClean, similarity, llmScore)

	return models.ScoreResult{
		TaskID:              run.TaskID,
		Agent:               run.Agent,
		TestsPass:           testsPass,
		TestOutput:          testOutput,
		LintClean:           lintClean,
		LintOutput:          lintOutput,
		DiffSimilarity:      similarity,
		LLMJudgeScore:       llmScore,
		LLMJudgeExplanation: llmExplanation,
		OverallScore:        overall,
	}
}

// DiffSimilarity computes the Jaccard similarity of changed lines between two diffs.
func DiffSimilarity(expected, actual string) float64 {
	exp := extractChangedLines(expected)
	act := extractChangedLines(actual)

	if len(exp) == 0 && len(act) == 0 {
		return 1.0
	}
	if len(exp) == 0 || len(act) == 0 {
		return 0.0
	}

	intersection := 0
	for line := range exp {
		if act[line] {
			intersection++
		}
	}
	union := len(exp) + len(act) - intersection
	return float64(intersection) / float64(union)
}

func extractChangedLines(diff string) map[string]bool {
	lines := map[string]bool{}
	for _, line := range strings.Split(diff, "\n") {
		if (strings.HasPrefix(line, "+") || strings.HasPrefix(line, "-")) &&
			!strings.HasPrefix(line, "+++") && !strings.HasPrefix(line, "---") {
			lines[strings.TrimSpace(line)] = true
		}
	}
	return lines
}

func (s *Scorer) runTests(run models.RunResult, task models.TaskDefinition, repoPath string) (*bool, string) {
	commands := task.TestCommands
	if len(commands) == 0 {
		commands = detectTestCommands(repoPath)
	}
	if len(commands) == 0 {
		return nil, ""
	}

	// Checkout base commit
	if out, err := exec.Command("git", "-C", repoPath, "checkout", task.BaseCommit).CombinedOutput(); err != nil {
		f := false
		return &f, fmt.Sprintf("git checkout failed: %s", out)
	}

	// Apply diff
	if run.ActualDiff != "" {
		applyCmd := exec.Command("git", "-C", repoPath, "apply", "--allow-empty")
		applyCmd.Stdin = strings.NewReader(run.ActualDiff)
		if out, err := applyCmd.CombinedOutput(); err != nil {
			f := false
			return &f, "failed to apply diff: " + string(out)
		}
	}

	var outputs []string
	allPass := true
	for _, cmd := range commands {
		c := exec.Command("sh", "-c", cmd)
		c.Dir = repoPath
		out, _ := c.CombinedOutput()
		outputs = append(outputs, string(out))
		if c.ProcessState == nil || c.ProcessState.ExitCode() != 0 {
			allPass = false
		}
	}
	return &allPass, strings.Join(outputs, "\n")
}

func (s *Scorer) runLint(task models.TaskDefinition, repoPath string) (*bool, string) {
	var changedFiles []string
	if files, ok := task.Metadata["files_changed"]; ok {
		switch v := files.(type) {
		case []string:
			changedFiles = v
		case []any:
			for _, f := range v {
				if str, ok := f.(string); ok {
					changedFiles = append(changedFiles, str)
				}
			}
		}
	}
	commands := detectLintCommands(repoPath, changedFiles)
	if len(commands) == 0 {
		return nil, ""
	}

	var outputs []string
	allClean := true
	for _, cmd := range commands {
		c := exec.Command("sh", "-c", cmd)
		c.Dir = repoPath
		out, _ := c.CombinedOutput()
		outputs = append(outputs, string(out))
		if c.ProcessState == nil || c.ProcessState.ExitCode() != 0 {
			allClean = false
		}
	}
	return &allClean, strings.Join(outputs, "\n")
}

func (s *Scorer) llmJudge(run models.RunResult, task models.TaskDefinition, expectedDiff string) (*int, string) {
	if s.NoLLM {
		return nil, ""
	}

	expDisplay := expectedDiff
	if expDisplay == "" {
		expDisplay = "(not available)"
	}
	actDisplay := run.ActualDiff
	if actDisplay == "" {
		actDisplay = "(empty — agent produced no changes)"
	}

	userContent := fmt.Sprintf(llmJudgePrompt, task.Prompt, expDisplay, actDisplay)

	msg, err := s.client.Messages.New(context.Background(), anthropic.MessageNewParams{
		Model:     anthropic.F(anthropic.Model(judgeModel)),
		MaxTokens: anthropic.F(int64(512)),
		Messages: anthropic.F([]anthropic.MessageParam{
			anthropic.NewUserMessage(anthropic.NewTextBlock(userContent)),
		}),
	})
	if err != nil {
		fmt.Printf("  LLM judge failed: %v\n", err)
		return nil, err.Error()
	}

	var text string
	for _, block := range msg.Content {
		if block.Type == anthropic.ContentBlockTypeText {
			text = block.Text
			break
		}
	}

	score, explanation := parseJudgeResponse(text)
	return score, explanation
}

var (
	scoreRe       = regexp.MustCompile(`(?m)^SCORE:\s*(\d+)`)
	explanationRe = regexp.MustCompile(`(?m)^EXPLANATION:\s*(.+)`)
)

func parseJudgeResponse(text string) (*int, string) {
	var score *int
	if m := scoreRe.FindStringSubmatch(text); m != nil {
		if n, err := strconv.Atoi(m[1]); err == nil {
			clamped := n
			if clamped < 1 {
				clamped = 1
			}
			if clamped > 5 {
				clamped = 5
			}
			score = &clamped
		}
	}
	explanation := ""
	if m := explanationRe.FindStringSubmatch(text); m != nil {
		explanation = strings.TrimSpace(m[1])
	}
	return score, explanation
}

func computeOverall(testsPass, lintClean *bool, diffSim float64, llmScore *int) float64 {
	totalWeight := 0.0
	weightedSum := 0.0

	if testsPass != nil {
		totalWeight += 0.35
		if *testsPass {
			weightedSum += 0.35
		}
	}
	if lintClean != nil {
		totalWeight += 0.10
		if *lintClean {
			weightedSum += 0.10
		}
	}
	totalWeight += 0.25
	weightedSum += 0.25 * diffSim

	if llmScore != nil {
		totalWeight += 0.30
		weightedSum += 0.30 * (float64(*llmScore) / 5.0)
	}

	if totalWeight == 0 {
		return 0
	}
	v := weightedSum / totalWeight
	// Round to 4 decimal places
	return float64(int(v*10000+0.5)) / 10000
}

func detectTestCommands(repoPath string) []string {
	if fileExists(filepath.Join(repoPath, "pytest.ini")) || fileExists(filepath.Join(repoPath, "pyproject.toml")) {
		return []string{"python -m pytest"}
	}
	if fileExists(filepath.Join(repoPath, "package.json")) {
		return []string{"npm test"}
	}
	if fileExists(filepath.Join(repoPath, "Makefile")) {
		return []string{"make test"}
	}
	return nil
}

func detectLintCommands(repoPath string, changedFiles []string) []string {
	var pyFiles, jsFiles []string
	for _, f := range changedFiles {
		if strings.HasSuffix(f, ".py") {
			pyFiles = append(pyFiles, f)
		} else if strings.HasSuffix(f, ".js") || strings.HasSuffix(f, ".ts") ||
			strings.HasSuffix(f, ".jsx") || strings.HasSuffix(f, ".tsx") {
			jsFiles = append(jsFiles, f)
		}
	}
	var cmds []string
	if len(pyFiles) > 0 {
		cmds = append(cmds, "ruff check "+strings.Join(pyFiles, " "))
	}
	if len(jsFiles) > 0 {
		cmds = append(cmds, "eslint "+strings.Join(jsFiles, " "))
	}
	return cmds
}

func fileExists(path string) bool {
	_, err := os.Stat(path)
	return err == nil
}
